### PR TITLE
btrfs-progs: Update to version 5.2.1

### PR DESCRIPTION
**Pull request by BKPepe**
Maintainer: @Cynerd 
Compile tested: Turris MOX, cortexa53, OpenWrt master
Run tested: Turris MOX, cortexa53, OpenWrt 19.07 (see #9646)

Description:
- GPL-2.0 is deprecated License Identifier according to SPDX.
- Correct it to GPL-2.0-or-later
- Update to version 5.2.1.

[Changes](https://git.kernel.org/pub/scm/linux/kernel/git/kdave/btrfs-progs.git/tree/CHANGES):
```
btrfs-progs-5.2.1 (2019-07-26)
  * scrub status: fix ETA calculation after resume
  * check: fix crash when using -Q
  * restore: fix symlink owner restoration
  * mkfs: fix regression with mixed block groups
  * core: fix commit to process all delayed refs
  * other:
    * minor cleanups
    * test updates

btrfs-progs-5.2 (2019-07-05)
  * subvol show: print qgroup information when available
  * scrub:
    * status: show ETA, revamp the whole output
    * fix reading/writing of last position on resume/cancel, potentially
      skipping part of the filesystem on next resume
  * dump-tree: add new option --noscan to use only devices given on the
    commandline
  * all-in-one binary (busybox style) with mkfs.btrfs, btrfs-image,
    btrfs-convert, btrfstune
  * image: fix hang when there are more than 32 cpus online and compression is
    requested
  * convert: fix some false ENOSPC errors when --rootdir is used
  * build: fix gcc9 warnings
  * core changes
    * command handling cleanups
    * dead code removal
    * cmds-* files moved to cmds/
    * other shared userspace files moved to common/
    * utils.c split into more files
    * preparatory work for more output formats
    * libbtrfsutil: fix unaligned access
  * other
    * new and updated tests
    * fix tests so CI passes again
    * sb-mod can modify more superblock items
```

### CODE REVIEW

**Comment by neheb** [comment]
The all in one binary change looks interesting. Can we switch to that?

I may or may not have had a hand in getting that in.

**Comment by neheb** [comment]
Another tip: Add USE_PROCD=0 to the init script. Some parties upstream want to default to procd for some reason.

**Comment by neheb** [comment]
License should be GPL-2.0-only according to the source files.

**Comment by Cynerd** [comment]
> The all in one binary change looks interesting. Can we switch to that?
> 
> I may or may not have had a hand in getting that in.

I would rather update it and later introduced that with separate commit. My I ask, when you had some hand in to it, what is an advantage? Is resulting binary much smaller than original sum of all binaries? I suspect that. I just want to be sure.



> Another tip: Add USE_PROCD=0 to the init script. Some parties upstream want to default to procd for some reason.

Second thing here is that I would like to move btrfs scan to preinit. I don't see why we should have it as init script when it does not make sense to (in theory) restart it. It is here just to trigger device scan at boot time. Any other time user can and will probably use `btrfs` command directly. I am of course open to discussions on this topic.

**Comment by neheb** [comment]
It should be smaller, yes. There's a lot of duplicate code among the various binaries.

No comment on the btrfs scan thing. I've never touched it.

**Comment by neheb** [comment]
Quick test:

From 2064740 bytes

To 1296258  bytes

**Comment by Cynerd** [comment]
@neheb do you have any idea why `btrfs-map-logical` and `btrfs-select-super` are not boxed in to single binary?

It kind of complicates stuff as there has to be not only box build but also regular build to build those tools.

**Comment by Cynerd** [comment]
@neheb would you review this? Update is done by @BKPepe and later changes are done by me so third party reviewer would be welcome.

File: utils/btrfs-progs/files/btrfs-scan.init [hunk 1]
```diff
@@ -1,9 +1,7 @@
-#!/bin/sh /etc/rc.common
-# Copyright (C) 2014 OpenWrt.org
+#!/bin/sh
 
-START=19
-
-start() {
-	grep -q btrfs /proc/filesystems && /usr/bin/btrfs device scan
+preinit_btrfs_scan() {
+	grep -vq btrfs /proc/filesystems || btrfs device scan
```
Comment (neheb) [inline utils/btrfs-progs/files/btrfs-scan.init]: So this is interesting :). You changed /usr/bin/btrfs to just btrfs. If I take a look at my OpenWrt device, ps returns mostly binaries with the prefix.

I don't know if there's a standard here that should be followed.

File: utils/btrfs-progs/Makefile [hunk 2]
```diff
@@ -72,8 +76,9 @@ define Package/btrfs-progs/install
 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libbtrfs.so* $(1)/usr/lib
 	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libbtrfsutil.so* $(1)/usr/lib
 	$(INSTALL_DIR) $(1)/usr/bin
-	$(INSTALL_BIN) $(addprefix $(PKG_INSTALL_DIR)/usr/bin/, $(progs)) $(1)/usr/bin/
-	$(LN) btrfs $(1)/usr/bin/btrfsck
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/btrfs.box $(1)/usr/bin/btrfs
+	$(foreach prog,$(boxprogs),$(LN) btrfs $(1)/usr/bin/$(prog);)
+	$(foreach prog,$(progs),$(INSTALL_BIN) $(PKG_BUILD_DIR)/$(prog) $(1)/usr/bin/;)
```
Comment (neheb) [inline utils/btrfs-progs/Makefile]: If you're not going to use the INSTALL_DIR anymore, might as well remove PKG_INSTALL.

File: utils/btrfs-progs/Makefile [hunk 3]
```diff
@@ -61,6 +61,10 @@ ifneq ($(CONFIG_BTRFS_PROGS_ZSTD),y)
 CONFIGURE_ARGS += --disable-zstd
 endif
 
+MAKE_INSTALL_FLAGS += BUILD_PROGRAMS=0
```
Comment (neheb) [inline utils/btrfs-progs/Makefile]: Would removing PKG_INSTALL accomplish the same thing?

File: utils/btrfs-progs/Makefile [hunk 4]
```diff
@@ -6,16 +6,16 @@
...
 
 PKG_SOURCE:=$(PKG_NAME)-v$(PKG_VERSION).tar.xz
 PKG_SOURCE_URL:=@KERNEL/linux/kernel/people/kdave/btrfs-progs
-PKG_HASH:=9cb91b7de9e10aa6bbf2b003f60bb3f5e5b1984a8008fad7c4b2d3978f5ebe1b
+PKG_HASH:=36ac4a0198ffff79d5800c537ea4b19769a8fd3ad870f75413d25b20e2d83233
 PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-v$(PKG_VERSION)
 
 PKG_MAINTAINER:=Karel Kočí <karel.koci@nic.cz>
-PKG_LICENSE:=GPL-2.0
+PKG_LICENSE:=GPL-2.0-or-later
```
Comment (neheb) [inline utils/btrfs-progs/Makefile]: Should be changed to -only. The source files don't mention version 2 or later. Only version 2.

**Comment by neheb** [comment]
From 2083265

To 1052107 on ARMEB. This is a very welcome improvement.

File: utils/btrfs-progs/Makefile (hunk 4 above)
Comment (BKPepe) [reply utils/btrfs-progs/Makefile]: [Alpine Linux](https://pkgs.alpinelinux.org/packages?name=btrfs-progs&branch=edge&repo=main) and got a little bit confused by [repology.org](https://repology.org/project/btrfs-progs/packages), but yeah. Most kernel stuff is under GPL-2.0-only. I will change it.

Comment (BKPepe) [reply utils/btrfs-progs/Makefile]: Fixed.

File: utils/btrfs-progs/Makefile (hunk 2 above)
Comment (neheb) [reply utils/btrfs-progs/Makefile]: Tested. This is needed for the libraries.

File: utils/btrfs-progs/Makefile (hunk 3 above)
Comment (neheb) [reply utils/btrfs-progs/Makefile]: Never mind.

**Pull request merged**